lib/docker_system.rb

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2013-2015 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of version 3 of the GNU General Public License as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+
+class DockerSystem < System
+  attr_accessor :image
+
+  def type
+    "docker"
+  end
+
+  def initialize(image)
+    @image = image
+    @container = DockerContainer.new(image)
+  end
+
+  def start
+    @container.start
+  end
+
+  def stop
+    @container.stop
+  end
+
+  def run_command(*args)
+    Machinery.logger.info("Running '#{args}'")
+
+    @container.run_command(*args)
+  end
+
+  def check_retrieve_files_dependencies
+    # Files are retrieved using the `docker cp` command, so there are no additional dependencies
+  end
+
+  def check_create_archive_dependencies
+    # Archives are created using the machinery-helper binary, so there are no additional
+    # dependencies
+  end
+
+  # Reads a file from the System. Returns nil if it does not exist.
+  def read_file(file)
+    run_command("cat", file, stdout: :capture)
+  rescue Cheetah::ExecutionFailed => e
+    if e.status.exitstatus == 1
+      # File not found, return nil
+      return
+    else
+      raise
+    end
+  end
+
+  # Removes a file from the System
+  def remove_file(file)
+    @container.run_command("rm", file)
+  rescue => e
+    raise Machinery::Errors::RemoveFileFailed.new(
+      "Could not remove file '#{file}'.\nError: #{e}"
+    )
+  end
+
+  # Copies a file to the system
+  def inject_file(source, destination)
+    @container.inject_file(source, destination)
+  end
+
+  # Retrieves files specified in file_list from the container
+  def retrieve_files(file_list, destination)
+    @container.retrieve_files(file_list, destination)
+  end
+
+  # Retrieves files specified in file_list from the container and creates an archive.
+  def create_archive(file_list, archive, exclude = [])
+    @container.create_archive(file_list, archive, exclude)
+  end
+
+  def requires_root?
+    false
+  end
+end
```

Comment:
I agree, but from the discussion we had I thought it was your idea to abstract away the actual docker access (which partly happens through the API and partly throught the docker command) from the DockerSystem, but maybe I misunderstood.

I'm perfectly fine with merging those two classes together.